### PR TITLE
Add Fedora CI research

### PR DESCRIPTION
**Pull request by mfocko**
Fixes packit/packit-service#2462

### CODE REVIEW

**Review by lbarcziova — APPROVED** [review]
thanks!

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:168]: IIRC, right now if there are no jobs configured on upstream, we have some implicit definitions (Copr build and TF).

The same could be done on downstream, but for Koji builds, and TF.

Comment (lbarcziova) [reply research/integrations/fedora-ci/index.md:168]: I see now, could you please add this explanation there?

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:40]: And we have scratch build functionality for upstream PRs available as well if this is anyhow relevant.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:34]: This is a very interesting point -- do we want to be consistent as much as possible with Packit's upstream implementation or be close to the real Fedora builds.

I agree the suggested approach `fedpkg srpm` makes more sense from the user perspective since this results in the production-like result.

Comment (lachmanfrantisek) [reply research/integrations/fedora-ci/index.md:34]: How do we want to handle sources? (Do we want to continue with the current trend of requiring sources in the lookaside cache?)

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:50]: Are some of these somehow dependant?

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:149]: Do you want to reuse the FedraCI-triggered Koji scratch build or trigger our own to be independent?

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:158]: Would this mean `dist-git CI` and the current Packit? Or, Packit upstream and Packit Fedora automation (including Koji, Bodhi and maybe also propose/pull)?

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:178]: :+1:

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:186]: That would be really awesome and close to the ?onGitLab approach. We might want to submit a Fedora change for this.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:205]: We can also reuse the Copr integration but being close to the final system might be better in this place.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:216]: We can ask TF if we shouldn't motivate people to go away by not supporting it...;)

Comment (lachmanfrantisek) [reply research/integrations/fedora-ci/index.md:216]: Is the presence detection of STI hard?

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:224]: Yes, manually-defined dependencies or reverse-dependencies are what people have been interested in so far.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:241]: So, use Packit config just as an enabler (for now)?

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:271]: Thanks for covering this one as well. Would be really good to count with these from the start.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:310]: Yes, but count on this possibility  (or global config ala zuul) since this should be the ultimate state.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:316]: Good point, you can link https://github.com/packit/packit-service/issues/1723

This also looks like another reason to have a separate instance.

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [inline research/integrations/fedora-ci/index.md:323]: Are these ordered? I would maybe put STI to the end or lower at least.

**Review by lachmanfrantisek — COMMENTED** [review]
Wow, you've included a lot of interesting subjects!

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:216]: > Is the presence detection of STI hard?

Have no clue :) I have **never** even seen the STI tests in the dist-git :D

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:34]: > This is a very interesting point -- do we want to be consistent as much as possible with Packit's upstream implementation or be close to the real Fedora builds.
> 
> I agree the suggested approach `fedpkg srpm` makes more sense from the user perspective since this results in the production-like result.

With respect to downstream, I would be definitely voting for avoiding adjusting the SRPM creation. Adjusting the SRPM creation in any way enables the chance of inconsistencies, i.e., I have green CI, but once I merge it doesn't build. I don't think many maintainers would like that :D

> How do we want to handle sources? (Do we want to continue with the current trend of requiring sources in the lookaside cache?)

I'm not really sure, but if we go deep into the possible improvements then it's quite possible that most of this initiative will not be achievable :)

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:40]: I thought we do non-scratch on upstream :eyes: I don't know how Koji handles “the heat”

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:50]: From the excerpt here (linting, installability, tests), it feels like linting should not be a hard blocker. I've seen the installability tests fail (even if tests passed) just because of the network flakes and such, so…

In the sense of saving the resources, it definitely makes sense to make _full test suite_ dependent on the _installability_, but if it's flakey, it's just annoying.

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:149]: :thinking: If Packit is to replace Fedora CI, there is no _Fedora CI-triggered Koji scratch build_ :P

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:158]: I've been thinking about a way to split this in some reasonable matter :D Based on the way Packit operates, the most reasonable would be splitting the upstream / downstream parts, i.e.,

- Packit Service (handles upstream **and** `propose-downstream` (as it is triggered by upstream))
- “Packit as Fedora CI” (handles downstream **and** `pull-from-upstream`)

**However** I rant somewhere about the config in the dist-git and default config :D and this complicates things a lot :/ Maybe we could treat PRs in a special way for the _downstream PRs_ (Koji + TF).

---

feels like whack-a-mole…

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:186]: With regards to your question about handling sources (somewhere above), it kinda means that the sources **must be** in the lookaside cache already, if we were to just tag the builds for testing/stable.

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:205]: We can use Koji scratch build… TBF I'm not sure why Zuul does it that way, but it could be caused by the fact that Zuul is not event-driven, i.e., it's not possible for it to wait for the build to finish and reuse the Koji build. :eyes: That doesn't make sense either… so it does both Koji scratch **and** its own build to give the Testing Farm?

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:241]: Maybe we could _abuse_ `packit_instances` to tell apart normal Packit config and what is supposed to be a Fedora CI? With one config for both Packit and “Packit as Fedora CI”, it will eventually become an incomprehensible mess (not to mention syncing of configs from upstream to downstream)

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:323]: no, they're not; will reorder

File: research/integrations/fedora-ci/index.md
Comment (mfocko) [reply research/integrations/fedora-ci/index.md:316]: done

File: research/integrations/fedora-ci/index.md
Comment (lachmanfrantisek) [reply research/integrations/fedora-ci/index.md:305]: Yes, on the one hand, people want as few configs as possible, but on the other hand, it can easily become a mess.

The idea with instances looks good.

Also, we already have Packit configs in quite a few projects that might be useful with some sane defaults..;) (To not need to go through the full onboarding again.)  Also, as you mentioned elsewhere, ideally not much confing would be needed. So, something to turn this on might be enough.

File: research/integrations/fedora-ci/index.md
Comment (lbarcziova) [reply research/integrations/fedora-ci/index.md:190]: + maybe also worth considering [this comment](https://github.com/packit/packit/issues/2414#issuecomment-2381490446)

**Comment by mfocko** [comment]
:rocket:

**Pull request merged**
